soap/MTOMEncoder.go

```diff
@@ -0,0 +1,250 @@
+	binaryFields := make([]reflect.Value, 0)
+	getBinaryFields(v, &binaryFields)
+	enableMTOMMode(binaryFields)
+
+	var partWriter io.Writer
+	var err error
+
+	h := make(textproto.MIMEHeader)
+	h.Set("Content-Type", `application/xop+xml`)
+	h.Set("Content-Transfer-Encoding", "8bit")
+
+	if partWriter, err = e.writer.CreatePart(h); err != nil {
+		return err
+	}
+	xmlEncoder := xml.NewEncoder(partWriter)
+	if err := xmlEncoder.Encode(v); err != nil {
+		return err
+	}
+
+	for _, fld := range binaryFields {
+		pkg := fld.Interface().(*Binary)
+		h := make(textproto.MIMEHeader)
+		if pkg.contentType == "" {
+			pkg.contentType = "application/octet-stream"
+		}
+		h.Set("Content-Type", pkg.contentType)
+		h.Set("Content-ID", fmt.Sprintf("<%s>", pkg.packageID))
+		h.Set("Content-Transfer-Encoding", "binary")
+		if partWriter, err = e.writer.CreatePart(h); err != nil {
+			return err
+		}
+		partWriter.Write(*pkg.content)
+	}
+
+	return nil
+}
+
+func (e *mtomEncoder) Flush() error {
+	return e.writer.Close()
+}
+
+func getBinaryFields(data interface{}, fields *[]reflect.Value) {
+	v := reflect.Indirect(reflect.ValueOf(data))
+
+	if v.Kind() != reflect.Struct {
+		return
+	}
+	for i := 0; i < v.NumField(); i++ {
+		if !v.Field(i).CanInterface() {
+			continue
+		}
+		f := v.Field(i)
+		if _, ok := f.Interface().(*Binary); ok {
+			*fields = append(*fields, f)
+		} else {
+			getBinaryFields(f.Interface(), fields)
+		}
+	}
+}
+
+func enableMTOMMode(fields []reflect.Value) {
+	for _, f := range fields {
+		b := f.Interface().(*Binary)
+		b.useMTOM = true
+	}
+}
+
+func (e *mtomEncoder) Boundary() string {
+	return e.writer.Boundary()
+}
+
+type mtomDecoder struct {
+	reader *multipart.Reader
+}
+
+func getMtomHeader(contentType string) (string, error) {
+	mediaType, params, err := mime.ParseMediaType(contentType)
+	if err != nil {
+		return "", err
+	}
+
+	if strings.HasPrefix(mediaType, "multipart/") {
+		boundary, ok := params["boundary"]
+		if !ok || boundary == "" {
+			return "", fmt.Errorf("Invalid multipart boundary: %s", boundary)
+		}
+
+		cType, ok := params["type"]
+		if !ok || cType != "application/xop+xml" {
+			// Process as normal xml (Don't resolve XOP parts)
+			return "", nil
+		}
+
+		startInfo, ok := params["start-info"]
+		if !ok || startInfo != "application/soap+xml" {
+			return "", fmt.Errorf(`Expected param start-info="application/soap+xml", got %s`, startInfo)
+		}
+		return boundary, nil
+	}
+
+	return "", nil
+}
+
+func newMtomDecoder(r io.Reader, boundary string) *mtomDecoder {
+	return &mtomDecoder{
+		reader: multipart.NewReader(r, boundary),
+	}
+}
+
+func (d *mtomDecoder) Decode(v interface{}) error {
+	fields := make([]reflect.Value, 0)
+	getBinaryFields(v, &fields)
+
+	packages := make(map[string]*Binary, 0)
+	for {
+		p, err := d.reader.NextPart()
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return err
+		}
+
+		contentType, ok := p.Header["Content-Type"]
```

Comment:
we will be safer here if we use https://golang.org/pkg/mime/#ParseMediaType instead, since `Content-Type` may come with a charset as well.

Reply:
oh, I saw you using it further up. We don't need to do it here and I'm missing something?

Reply:
The first ParseMediaType is needed to detect if the message as a whole is using XOP encoding.
The Content-Type in the Decode is used to determine the content on the individual MIME parts (i.e. application/xop+xml vs binary parts).
BTW I changed `p.Header[]` lines with `p.Header.Get` in order to safely get the first element of the content type.